interp/src/primitives/stateful.rs

```diff
@@ -51,6 +56,37 @@ impl StdMultPipe {
 }
 
 impl Primitive for StdMultPipe {
+    fn do_tick(&mut self) -> Vec<(ir::Id, OutputValue)> {
+        if let Some(Some(out)) = self.queue.pop_back() {
+            //push update to the front
+            self.queue.push_front(self.update.take());
+            //assert queue still has length 3
+            assert_eq!(
+                self.queue.len(),
+                3,
+                "std_mult_pipe's internal queue has length {} != 3",
+                self.queue.len()
+            );
+            //return vec w/ out and done
+            vec![
+                (ir::Id::from("out"), out.into()),
+                (ir::Id::from("done"), Value::bit_high().into()),
+            ]
+        } else {
+            //push update to the front
+            self.queue.push_front(self.update.take());
+            //assert queue still has length 3
+            assert_eq!(
+                self.queue.len(),
+                3,
+                "std_mult_pipe's internal queue has length {} != 3",
+                self.queue.len()
+            );
+            //return empty vec
+            vec![]
+        }
+    }
+    //
```

Comment:
Correct me if I'm wrong, but as implemented it seems to me that this takes 4 cycles to produce the first output, instead of three.


TIME 0
  update: VAL
  buffer: [ none, none, none ]

TIME 1 
  update: none
  buffer: [ VAL, none, none]

TIME 2
  update: none
  buffer: [ none, VAL, none]

TIME 3
  update: none
  buffer: [ none, none, VAL]

TIME 4
  update: none
  buffer: [ none, none, none]  => Output VAL

Seems like 4 do_ticks must occur here, no?

Reply:
Yea, you're right. Not sure how I missed that b/c I wrote the tests... I'll change it